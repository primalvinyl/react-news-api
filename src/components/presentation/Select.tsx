import React from 'react';
import styles from './Select.module.scss';

const Select = React.forwardRef<HTMLSelectElement, SelectType>(
    (props, ref): React.ReactElement => {

    const {
        label,
        id,
        name,
        className,
        value,
        options,
        errors,
        touched,
        required,
        disabled,
        handleChange,
        handleBlur
    } = props;

    return (
        <div className={styles.root}>
            {label && (<label htmlFor={id} className={styles.formLabel}>{label}</label>)}
            <div className={styles.formControlWrapper}>
                <select
                    className={className}
                    name={name}
                    id={id}
                    value={value}
                    ref={ref}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    required={required}
                    disabled={disabled}
                >
                    {options.map(option => (
                        <option key={option.value} value={option.value}>
                            {option.display}
                        </option>
                    ))}
                </select>
                {(errors && touched) && (errors[id] && touched[id]) && (
                    <div className={styles.formFieldError}>
                        {errors[id]}
                    </div>
                )}
            </div>
        </div>
    );
});

type SelectType = {
    readonly id: string;
    readonly value: string;
    readonly options: {value: string; display: string}[];
    readonly className?: string;
    readonly name?: string;
    readonly label?: string;
    readonly required?: boolean;
    readonly disabled?: boolean;
    readonly errors?: { [key: string]: string };
    readonly touched?: { [key: string]: string };
    readonly handleChange?: (event: React.ChangeEvent<HTMLSelectElement>) => void;
    readonly handleBlur?: (event: React.FocusEvent<HTMLSelectElement>) => void;
};

Select.defaultProps = {
    className: undefined,
    name: undefined,
    label: undefined,
    required: false,
    disabled: false,
    errors: {},
    touched: {},
    handleChange: (): void => {},
    handleBlur: (): void => {},
}

export default Select;
