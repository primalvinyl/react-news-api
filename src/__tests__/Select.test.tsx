import * as React from 'react';
import { shallow } from 'enzyme';
import Renderer from 'react-test-renderer';
import Select from '../components/presentation/Select';

describe('Select', () => {
    it('renders', () => {
        const myWrapper = shallow(
            <Select
                id="test"
                value="test"
                options={[]} />
        );
        expect(myWrapper.exists());
    });
    it('matches snapshot', () => {
        const component = Renderer.create(
            <Select
                id="test"
                value="test"
                options={[]} />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});