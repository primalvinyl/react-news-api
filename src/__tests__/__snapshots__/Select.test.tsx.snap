// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select matches snapshot 1`] = `
<div
  className="root"
>
  <div
    className="formControlWrapper"
  >
    <select
      disabled={false}
      id="test"
      onBlur={[Function]}
      onChange={[Function]}
      required={false}
      value="test"
    />
  </div>
</div>
`;
